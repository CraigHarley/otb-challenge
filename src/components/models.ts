import {IHotel} from "../types";

export const hotels: IHotel[] = [
    {
        adults: 2,
        children: 0,
        infants: 1,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£3,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Iberostar Grand Salome",
        image: "hotel-image-1.png",
        stars: 4,
        date: new Date('2019-07-03')
    },
    {
        adults: 2,
        children: 1,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£1,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Wow - so fancy",
        image: "hotel-image-2.png",
        stars: 2,
        date: new Date('2019-07-03')
    },
    {
        adults: 1,
        children: 0,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£5,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Swank-ville",
        image: "hotel-image-3.png",
        stars: 4,
        date: new Date('2019-07-03')
    },
    {
        adults: 2,
        children: 0,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£2,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Really Fancy Hotel",
        image: "hotel-image-1.png",
        stars: 5,
        date: new Date('2019-07-03')
    }
]
