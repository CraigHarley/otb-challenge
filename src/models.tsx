import React from 'react';

import {IHotel, ISort} from "./types";
import {SortTypes} from "./types/enums";

export const sorts: ISort[] = [
    {
        name: 'alphabetically',
        type: SortTypes.ALPHA,
        func: new Function(), // todo
        icon: <i className="float-right fas fa-sort-alpha-down"/>
    },

    {
        name: 'price',
        type: SortTypes.PRICE,
        func: new Function(), // todo
        icon: <i className="float-right fas fa-pound-sign"/>
    },

    {
        name: 'star rating',
        type: SortTypes.STARS,
        func: new Function(), // todo
        icon: <i className="float-right fas fa-star"/>
    },
]

export const hotels: IHotel[] = [
    {
        adults: 2,
        children: 0,
        infants: 1,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£3,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Iberostar Grand Salome",
        image: "hotel-image-1.png",
        stars: 4,
        date: new Date('2019-07-03')
    },
    {
        adults: 2,
        children: 1,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£1,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Wow - so fancy",
        image: "hotel-image-2.png",
        stars: 2,
        date: new Date('2019-07-03')
    },
    {
        adults: 1,
        children: 0,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£5,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Swank-ville",
        image: "hotel-image-3.png",
        stars: 4,
        date: new Date('2019-07-03')
    },
    {
        adults: 2,
        children: 0,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£2,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Really Fancy Hotel",
        image: "hotel-image-1.png",
        stars: 5,
        date: new Date('2019-07-03')
    }
]
