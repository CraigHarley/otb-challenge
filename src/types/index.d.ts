export interface IHotel {
    name: string;
    region: string;
    stars: number;
    price: string;
    adults: number;
    children: number;
    infants: number;
    info: string;
    image: string;
    date: Date;
}

export interface ISort {
    description: string;
    func: Function;
    icon: string;
}

export interface IListable {
    key?: number;
}
