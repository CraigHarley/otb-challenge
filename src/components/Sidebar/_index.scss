@import "src/variables";

.sidebar {
  grid-area: Sidebar;
  .button-container {
    min-width: 200px;
    max-width: 200px;
    margin: 2rem 5rem;
    button:first-child {
      border-top: none;
    }
    button {
      text-transform: lowercase;
      text-align: left;
      appearance: none;
      text-decoration: none;
      cursor: pointer;
      border: none;
      height: 3rem;
      min-width: 100%;
      color: $primary;
      border-top: 1px solid $primary;
      padding: 0.5rem 1rem 0.5rem 1rem;
      line-height: 1rem;
      font-size: small;

      &.is-active {
        background-color: $primary;
        color: white;

        &:hover {
          background-color: lighten($primary, 10%);

        }
      }
      &:hover {
        background-color: lighten($primary, 50%);

      }
    }
  }
}
