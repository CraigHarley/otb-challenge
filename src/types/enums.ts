export enum SortTypes {
    ALPHA,
    PRICE,
    STARS
}

export enum SortDirections {
    UP = 1,
    DOWN = -1,
}
