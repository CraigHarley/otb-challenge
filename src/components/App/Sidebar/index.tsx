import React, {ReactElement} from 'react';
import {css, StyleSheet} from 'aphrodite';

import {ISort} from '../../../types';
import {SortTypes} from '../../../types/enums';
import config from '../../../config'

export interface ISidebarProps {
    setChosenSort: (sort: SortTypes) => void;
    chosenSort: SortTypes;
    sorts: ISort[];
}

const styles = StyleSheet.create({
    buttonContainer: {
        minWidth: '200px',
        maxWidth: '200px',
        margin: '2rem 5rem'
    },
    button: {
        textTransform: 'lowercase',
        textAlign: 'left',
        appearance: 'none',
        textDecoration: 'none',
        cursor: 'pointer',
        border: 'none',
        height: '3rem',
        minWidth: '100%',
        color: config.styles.colors.primary,
        backgroundColor: 'whitesmoke',
        borderTop: `1px solid ${config.styles.colors.primary}`,
        padding: '0.5rem 1rem 0.5rem 1rem',
        lineHeight: '1rem',
        fontSize: 'small',
        ':hover': {
            backgroundColor: config.styles.colors.secondary,
            color: config.styles.colors.primary,
        }
    },
    firstButton: {
        borderTop: 'none'
    },
    isActive: {
        backgroundColor: config.styles.colors.primary,
        color: 'whitesmoke',
        ':hover': {
            backgroundColor: config.styles.colors.primary,
            color: 'whitesmoke',
        }
    }
});

const Sidebar = ({sorts, chosenSort, setChosenSort}: ISidebarProps): ReactElement => (
    <aside>
        <div className={css(styles.buttonContainer)}>
            {
                sorts.map(
                    ({name, icon, type}, key) => (
                        <button
                            name={`btn-sort-${key}`}
                            key={key}
                            className={
                                css(
                                    styles.button,
                                    (chosenSort === type) && styles.isActive,
                                    key === 0 && styles.firstButton
                                )
                            }
                            onClick={() => setChosenSort(type)}
                        >
                            Sort <strong>{name}</strong> {icon}
                        </button>
                    )
                )
            }
        </div>
    </aside>
);

export default Sidebar;
