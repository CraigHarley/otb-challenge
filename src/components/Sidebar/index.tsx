import React from 'react';
import {ISort} from "../../types";
import {SortTypes} from "../../types/enums";

export interface ISidebarProps {
    setChosenSort: (sort: SortTypes) => void;
    chosenSort: SortTypes;
    sorts: ISort[];
}

export default ({sorts, chosenSort, setChosenSort}: ISidebarProps): React.ReactNode => (
    <aside className="sidebar">
        <div className="button-container">
            {
                sorts.map(
                    ({name, icon, type}, key) => (
                        <button
                            key={key}
                            className={(chosenSort === type) ? 'is-active' : undefined}
                            onClick={() => setChosenSort(type)}
                        >
                            Sort <strong>{name}</strong> {icon}
                        </button>
                    )
                )
            }
        </div>
    </aside>
);
