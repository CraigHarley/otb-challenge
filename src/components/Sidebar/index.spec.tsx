import React from 'react';
import {mount} from 'enzyme';

import Sidebar from ".";
import {SortTypes} from "../../types/enums";
import {sorts} from "../../models";

const defaultProps = {
    setChosenSort: () => ({}),
    chosenSort: SortTypes.STARS,
    sorts
}

test('it should render with sorts',
    () =>
        expect(
            <Sidebar
                {...defaultProps}
            />
        )
            .toMatchSnapshot('default')
);

test('it should render without sorts',
    () =>
        expect(
            <Sidebar
                {
                    ...{
                        ...defaultProps,
                        sorts: []
                    }
                }
            />
        )
            .toMatchSnapshot('default')
);

test('when a button is clicked, it should call setChosenSort',
    () => {
        const setChosenSort = jest.fn();
        const element = mount(
            <Sidebar
                {
                    ...{
                        ...defaultProps,
                        setChosenSort
                    }
                }
            />
        );
        element.find('button').first().simulate('click');
        expect(setChosenSort).toHaveBeenCalled();
    }
);
