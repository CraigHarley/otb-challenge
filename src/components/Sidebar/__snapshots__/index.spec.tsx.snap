// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar it should render with sorts: default 1`] = `
.buttonContainer {
  min-width: 200px !important;
  max-width: 200px !important;
  margin: 2rem 5rem !important;
}

.button-firstButton {
  text-transform: lowercase !important;
  text-align: left !important;
  -webkit-appearance: none !important;
  -moz-appearance: none !important;
  appearance: none !important;
  text-decoration: none !important;
  cursor: pointer !important;
  border: none !important;
  height: 3rem !important;
  min-width: 100% !important;
  color: rgb(21, 47, 132) !important;
  background-color: whitesmoke !important;
  padding: 0.5rem 1rem 0.5rem 1rem !important;
  line-height: 1rem !important;
  font-size: small !important;
  border-top: none !important;
}

.button-firstButton:hover {
  background-color: rgb(255, 220, 0) !important;
  color: rgb(21, 47, 132) !important;
}

.button {
  text-transform: lowercase !important;
  text-align: left !important;
  -webkit-appearance: none !important;
  -moz-appearance: none !important;
  appearance: none !important;
  text-decoration: none !important;
  cursor: pointer !important;
  border: none !important;
  height: 3rem !important;
  min-width: 100% !important;
  color: rgb(21, 47, 132) !important;
  background-color: whitesmoke !important;
  border-top: 1px solid rgb(21, 47, 132) !important;
  padding: 0.5rem 1rem 0.5rem 1rem !important;
  line-height: 1rem !important;
  font-size: small !important;
}

.button:hover {
  background-color: rgb(255, 220, 0) !important;
  color: rgb(21, 47, 132) !important;
}

.button-o_O-isActive_zojkvj {
  text-transform: lowercase !important;
  text-align: left !important;
  -webkit-appearance: none !important;
  -moz-appearance: none !important;
  appearance: none !important;
  text-decoration: none !important;
  cursor: pointer !important;
  border: none !important;
  height: 3rem !important;
  min-width: 100% !important;
  border-top: 1px solid rgb(21, 47, 132) !important;
  padding: 0.5rem 1rem 0.5rem 1rem !important;
  line-height: 1rem !important;
  font-size: small !important;
  background-color: rgb(21, 47, 132) !important;
  color: whitesmoke !important;
}

.button-o_O-isActive_zojkvj:hover {
  background-color: rgb(21, 47, 132) !important;
  color: whitesmoke !important;
}

<aside>
  <div
    class="buttonContainer"
  >
    <button
      class="button-firstButton"
      name="btn-sort-0"
    >
      Sort 
      <strong>
        alphabetically
      </strong>
       
      <i
        class="float-right fas fa-sort-alpha-down"
      />
    </button>
    <button
      class="button"
      name="btn-sort-1"
    >
      Sort 
      <strong>
        by price
      </strong>
       
      <i
        class="float-right fas fa-pound-sign"
      />
    </button>
    <button
      class="button-o_O-isActive_zojkvj"
      name="btn-sort-2"
    >
      Sort 
      <strong>
        by star rating
      </strong>
       
      <i
        class="float-right fas fa-star"
      />
    </button>
  </div>
</aside>
`;

exports[`Sidebar it should render without sorts: default 1`] = `
.buttonContainer {
  min-width: 200px !important;
  max-width: 200px !important;
  margin: 2rem 5rem !important;
}

<aside>
  <div
    class="buttonContainer"
  />
</aside>
`;
